File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* currNode, int maxval, int minval) {
        if (currNode == nullptr) return maxval-minval;
        maxval = max(currNode->val, maxval);
        minval = min(currNode->val, minval);

        int left_diff = dfs(currNode->left, maxval, minval);
        int right_diff = dfs(currNode->right, maxval, minval);

        return left_diff > right_diff ? left_diff : right_diff;
    }

    int maxAncestorDiff(TreeNode* root) {
        int l = dfs(root->left, root->val, root->val);
        int r = dfs(root->right, root->val, root->val);

        return l > r ? l : r;
    }
};

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def maxAncestorDiff(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        def dfs(currNode, max_val, min_val):
            if not currNode:
                return max_val - min_val
            if currNode.val > max_val:
                max_val = currNode.val
            if currNode.val < min_val:
                min_val = currNode.val
            left = dfs(currNode.left, max_val, min_val)
            right = dfs(currNode.right, max_val, min_val)
            return left if left > right else right
        
        left_result = dfs(root.left, root.val, root.val)
        right_result = dfs(root.right, root.val, root.val)

        return left_result if left_result > right_result else right_result